actionpack/lib/action_dispatch/system_test_case.rb

```diff
@@ -166,7 +166,10 @@ def url_helpers
               include ActionDispatch.test_app.routes.mounted_helpers
 
               def url_options
-                default_url_options.reverse_merge(host: Capybara.app_host || Capybara.current_session.server_url)
+                test_url_options = ActionDispatch.test_app.routes.default_url_options
+                  .merge(ApplicationController.new.default_url_options)
```

Comment:
I'm not sure about this one. `ApplicationController` is not a required class in any application, so I don't think we should rely on it. Maybe we should add a configuration somewhere which default to `"ApplicationController"`?

Reply:
Good catch. To avoid merging with `nil` when `ApplicationController` is not defined, we could use something like this:

```ruby
test_url_options = ActionDispatch.test_app.routes.default_url_options
  .merge('ApplicationController'.safe_constantize&.new&.default_url_options || {})
```

Not very elegant though. Maybe you have a suggestion where a configuration which defaults to `ApplicationController`  could be added.

Reply:
We can go with your code right now and add the config later if necessary.

Reply:
All right, just updated this branch accordingly